src/rand_r.rs

```diff
@@ -0,0 +1,52 @@
+//! Rust implementation of C library function `rand_r`
+//!
+//! Licensed under the Blue Oak Model Licence 1.0.0
+use core::ffi::{c_int, c_uint};
+
+#[cfg_attr(not(feature = "rand_r"), export_name = "tinyrlibc_RAND_MAX")]
+#[cfg_attr(feature = "rand_r", no_mangle)]
+pub static RAND_MAX: c_int = 0x7FFF_FFFC as _;
```

Comment:
I tested all `2^32` possible seeds and the value never went over `0x7FFF_FFFC`. This seemed strange to me and I didn't really know what to do with `RAND_MAX` in that case. The original algorithm goes to up `0x7FFF_FFFD`, but does not include `0`. That is why I subtracted `1`, because I thought it would be way more unexpected to never get `0` then to never get `0x7FFF_FFFD`.

Reply:
How many times did you call rand()? The 10 or 11 bits that get pulled should be relatively uniform. It might be worth logging them individually. Then if they are uniform, you should get values from 0 to 2^31 - 1 by combining 31 random bits together.

Reply:
OK, I did some digging on this one.

1) Turns out `rand()` isn't required to be thread safe. So we don't actually need portable-atomic - a load/store pair is fine. It'll be no worse than a C implementation.

2) Also, newlib sets RAND_MAX to INT_MAX - and handles both 16-bit and 32-bit ints. I propose we do the same.

3) The algorithm proposed produces what look like uniformly random bits
  * To confirm, I took 15 bits from the top 16 bits, sampled 10 million times, and every value from 0 to 32767 was produced roughly the same number of times, with no missing values.
  * Therefore it should be fine do do that three times and combine the results into a random value from `0..(1<<31)-1`.